versioning/worker-versions.md

```diff
@@ -0,0 +1,97 @@
+
+Temporal workflow authors experience a number of difficulties when it comes to
+changing workflow code. The current techniques for versioning workflows are:
+
+**GetVersion & patch APIs**
+These APIs allow branching within the same workflow code to handle the presence
+or absence of a certain change. See them for [TS](https://docs.temporal.io/docs/typescript/patching#typescript-sdk-patching-api), [Java](https://docs.temporal.io/docs/java/versioning/), and [Go](https://docs.temporal.io/docs/go/versioning/).
+
+These APIs are confusing and challenging to use correctly, and clutter up code.
+We advise users to generally use them only to fix bugs that would otherwise prevent
+the workflow from completing.
+
+**Versioned task queues**
+Conceptually simpler, but operationally more effort, is to version task queues.
+IE: A worker operating on task queue `foo-v1` is changed, and when redeployed
+points at `foo-v2`.
+
+This avoids complicating the workflow code itself, but means users need to be
+aware of when all workflows on `foo-v1` are complete, allowing them to decomission
+the old workers. It also means potentially deploying larger numbers of workers.
+
+It also requires updating clients who are starting workflows to point at the
+newer task queue name.
+
+**Versioned workflow names**
+This approach is much like versioning task queue names, but trades off some of
+the operational complexity for some code maintenance complexity.
+
+Rather than needing to keep old workers around, users now need to keep the
+old code around until no more workflows using the old type are open.
+
+They still have the same problem with needing to tell clients to use the new
+workflow name.
+
+## Improving the situation
+
+The `patch` type APIs can't be elimiated completely. There will always be 
+situations where a "broken" workflow needs to be fixed to make progress. Such
+fixes are typically less common than just making changes to workflow logic, though.
+
+To make the more common case easier, we will provide a set of features that
+make the "versioned task queues" approach much easier to use.
+
+At a high level:
+* Provide a simple API for versioning entire workers (this proposal)
+* Allow clients to point at task queue `foo`, and automatically route new workflow executions
+  to the latest version of the `foo` queue.
+* Provide good notifications and information about when old workers may be
+  decomissioned
+
+Building on this base set of features, we plan to eventually implement
+automatic safe-rollout of changes, and possibly tools like k8s operators
+which can automatically manage worker deployments.
+
+## SDK API changes
+
+The meat of this proposal is simple:
+**Add a version field to worker options in each SDK**
+
+Because version information will need to be sent to the Temporal Cluster,
+we define it as a protobuf message (which will show up in our API somewhere):
+```protobuf
+message WorkerVersions {
+    string worker_version = 1;
```

Comment:
I'm kind of allergic to stuffing structured data into strings.. if we're actually going to enforce semver-like semantics on these, maybe this should be two or three ints instead?

Reply:
Great idea, will incorporate.

Reply:
Maybe the same for the SDKs. Meh, we may not need to expose a typed version per se since it may discourage setting it, but we should at least eagerly validate it at runtime (which of course we have to if putting the individual pieces into proto fields).

Reply:
May want to have the SDKs have a typed version too, though not required since they'll at least have to parse to fill this proto.

Reply:
Semver is more than just 3 ints, I'm also for strong types and TBH I'm not convinced this should be semver at all since we don't use the patch version or any other info semver carries (like build info / rc).

Reply:
We've been talking about `foo.bar` version, where `foo` is a backward-incompat change to workflow interface or impl, and `bar` is back-compat change to workflow impl, including `patch` calls. Why do we need the latter? / what do we do with it?

> **What if Tommy wants to make a patch?**
> Same process, except he can use the `patch` API if needed, and increments the minor version number.

Reply:
Fix bugs. 

EX: Your workflow code derefs a null pointer while parsing some data returned by an activity and your open WFs are stuck failing a wf task at that point.

To fix this, you might not _need_ to use `patch`, but say you wanted to remove the activity call entirely to fix this, so you use patch and in the branch w/ the patch you don't call the activity any more.

Reply:
> Semver is more than just 3 ints, I'm also for strong types and TBH I'm not convinced this should be semver at all since we don't use the patch version or any other info semver carries (like build info / rc).

(see my latest change)

True, and true, but the good news is users aren't sending this API directly. So, they can avoid caring about this, except I think we should ask the strings they write be semver-compatible (and we will then parse them). 99.99% of the time they can just write "4.2" and that'll work, but if they want to write "4.2.3-prerelease+build" then they can too, and we'll only use major/minor but we can show their whole string in UI/queries/whatever.

Reply:
I get that the design is the user increments the minor version to fix bugs, but what do we do with that information? Like with the major number, we change the task queue shadow version.

Reply:
Same thing - new workflows (and new workflow tasks within the same major version) are started using the most recent version, including the patch number.

Reply:
I'm not sure we want to let users specify full semver.
Would they be surprised if they bumped an rc or a patch and we simply ignored that?

Reply:
> Same thing - new workflows (and new workflow tasks within the same major version) are started using the most recent version, including the patch number.

Is the only benefit to this that when I deploy `2.2` and I leave my `2.0/2.1` workers up, new `2.*` tasks & executions will no longer go to `2.0/2.1` workers? 

If that's the only benefit, then I'd much prefer the current process—when I deploy a back-compat change to code, I send shutdown signal to current workers. (With major versioning, I'd send shutdown to just the existing v2 workers.) Then I don't have to change a version number every time I update my code. And also the API gets simpler—instead of "foo.bar" where I have to explain A) it's not semver, B) when to use which number, I just have one integer `workerVersion: 2`. Or could name it more informative, like:

```
backwardIncompatibleVersion
breakingWorkflowHistoryVersion
breakingHistoryVersion
breakingVersion
```

Reply:
@bergundy 
>  Would they be surprised if they bumped an rc or a patch and we simply ignored that?

I don't think they would. They have to understand the incrementing rules to even make use of this feature at all in the first place. If you understand them, then you understand we only care about major/minor, and anything else is just your personal bookkeeping. 

@lorensr 
So nothing's preventing you from doing it that way still if you think it's easier, just deploy and don't actually bump the minor number. New tasks would still work with the previous minor versions if you do bump. They just won't be routed to them, because presumably you were upgrading for a reason and would prefer to run on the new code.

The version _is_ a semver version. It adheres to the spec, we just happen to only match on / care about major and minor. It's exactly like specifying `^1.0` in npm land, which is a common thing to do anyway.

I suppose just one number is in some sense simpler, but, you still have to understand what constitutes a breaking change and what doesn't which seems to me to be most of the work.

Reply:
Gotcha. So we could support both ways. And we want to support the including-minor-version way because we expect there will be users who will be leaving up their `2.1` workers for some period of time after deploying `2.2`?

If yes, I'm wondering whether we should recommend / explain first what we think is the better DX path in the docs. Like:

- Breaking change: set `workerVersion: '2'`
- Not breaking:
  - If you can send shutdown signal to current v2 workers at the same time you deploy, do that
  - Otherwise, set `workerVersion: '2.1'` and deploy

Reply:
Any users who need to maintain availability, we'd expect there is a small period of overlap between 2.1 and 2.2. They need to avoid a stop-the-world-then-restart it situation which would harm availability while that happens.

Using 2.1 and 2.2 explicitly ensures that new tasks go to the new workers, instead of some new tasks potentially still taking the old path. Then once all open workflows have processed at least one task on 2.2, you're notified you know it's OK to retire the 2.1 workers.

Otherwise you could turn on all your next version workers, then once they're up, send the shutdown signal to your prev version ones, but while you are doing that it's likely that some new tasks will go to old workers. Probably most of the time, that's totally fine, but having the sequencing be more explicit and visible is a nice property. It also allows you to understand what workers have what versions without you needing to remember that yourself.

So I'm inclined to recommend that people use the minor version, but we can certainly point out that strictly speaking you don't have to use it if you don't want in the docs.

Reply:
> I don't think they would. They have to understand the incrementing rules to even make use of this feature at all in the first place. If you understand them, then you understand we only care about major/minor, and anything else is just your personal bookkeeping.

I don't think this is good DX, it basically means that either:
1. Our system doesn't distinguish between these 2 different workers and workflows may be routed to either.
2. Worker is rejected when it connects to the system because it uses a pre-registered version and a different binary checksum.

Reply:
It means 1. Are you suggesting it's more clear to explicitly only pass major/minor? I'm not sure that really fixes much. You still have to understand what they mean - and if you only allow that, users have no way to attach something they might want to show up when querying worker versions/etc like a build id. You could make a separate field for that, but then it's the same thing, you could say they might be confused why that field doesn't affect routing. Additionally, you can make the same argument about even that - if the user is just plowing ahead blindly they might expect incrementing minor to do something different than it does, which is sorta the same as having some incorrect expectation about incrementing patch and having that do something.

With the semver string we have:
* It's familiar to users
* Allows inclusion of extra metadata user might care about but system doesn't
* Future proof if we _do_ care about patch/etc later on
* Easy to write
* No change in routing with patch/prerelease/build is at least congruent with reasonable expectations about how they would behave 

TLDR: I don't think there's a version of this API that allows you to never understand anything and use it blindly, successfully, and the string has some other nice advantages

Reply:
Okay, I'm convinced this is the right approach